File: Piko/Sources/SFSketch.m
//
//  SFSketch.m
//
//  Copyright (c) 2014 Satoshi Muraki. All rights reserved.
//

#import "SFSketch.h"

@interface SFSketch ()

@property (nonatomic, readonly) NSMutableArray *nodes;

@end

@implementation SFSketch

- (instancetype)init
{
    self = [super init];
    if (self != nil) {
        _outputNode = [[SFOutputNode alloc] init];
        _nodes = [[NSMutableArray alloc] initWithCapacity:0];
    }
    return self;
}

- (void)appendNode:(SFNode *)node
{
    [_nodes addObject:node];
    [_delegate sketch:self didAppendNode:node];
}

- (void)removeNode:(SFNode *)node
{
    NSUInteger index = [_nodes indexOfObject:node];
    if (index == NSNotFound) {
        return;
    }

    for (SFInput *input in node.inputs) {
        SFOutput *output = input.source;
        if (output != nil) {
            [self disconnectOutput:output];
        }
    }

    for (SFOutput *output in node.outputs) {
        [self disconnectOutput:output];
    }

    [_nodes removeObjectAtIndex:index];
    [_delegate sketch:self didRemoveNode:node];
}

- (void)connect:(SFOutput *)src to:(SFInput *)dst
{
    SFOutput *currentSrc = dst.source;
    if (currentSrc != nil) {
        if (currentSrc == src) {
            return;
        } else {
            [self disconnectOutput:currentSrc];
        }
    }

    SFInput *currentDst = src.destination;
    if (currentDst != nil) {
        [self disconnectOutput:src];
    }

    src.destination = dst;
    dst.source = src;
}

- (void)disconnectOutput:(SFOutput *)output
{
    SFInput *input = output.destination;
    if (input != nil) {
        output.destination = nil;
        input.source = nil;
    }
}

@end

File: Piko/Sources/SFSketch.swift
//
//  SFSketch.swift
//
//  Copyright (c) 2014 Satoshi Muraki. All rights reserved.
//

import Foundation

@objc protocol SFSketchDelegate {
    func sketch(sketch: SFSketch, didAppendNode node: SFNode)
    func sketch(sketch: SFSketch, didRemoveNode node: SFNode)
}

@objc class SFSketch: NSObject {

    let outputNode: SFOutputNode
    var nodes: [SFNode]
    weak var delegate: SFSketchDelegate?

    override init() {
        outputNode = SFOutputNode()
        nodes = [outputNode]
        super.init()
    }

    func appendNode(node: SFNode) {
        nodes.append(node)
        delegate?.sketch(self, didAppendNode: node)
    }

    func removeNode(node: SFNode) {
        let index = nodes.indexOfObject(node)
        if let i = index {
            nodes.removeAtIndex(i)
            delegate?.sketch(self, didRemoveNode: node)
        }
    }

    func connect(src: SFOutput, to dst: SFInput) {

        if let currentSrc = dst.source {
            if currentSrc == src {
                return
            } else {
                self.disconnectOutput(currentSrc)
            }
        }

        if let oldDst = src.destination {
            self.disconnectOutput(src)
        }

        src.destination = dst;
        dst.source = src;
    }

    func disconnectOutput(output: SFOutput) {
        if let input = output.destination {
            output.destination = nil
            input.source = nil
        } else {
            assert(false)
        }
    }
}
